planner/core/preprocess.go

```diff
@@ -1685,3 +1697,103 @@ func (p *preprocessor) hasAutoConvertWarning(colDef *ast.ColumnDef) bool {
 	}
 	return false
 }
+
+func tryLockMDLAndUpdateSchemaIfNecessary(sctx sessionctx.Context, dbName model.CIStr, tbl table.Table, is infoschema.InfoSchema) (table.Table, error) {
+	if !variable.EnableMDL.Load() {
+		return tbl, nil
+	}
+	if is.SchemaMetaVersion() == 0 {
+		return tbl, nil
+	}
+	skipLock := false
+	if sctx.GetSessionVars().SnapshotInfoschema != nil {
+		return tbl, nil
+	}
+	if sctx.GetSessionVars().TxnCtx.IsStaleness {
+		return tbl, nil
+	}
+	if tbl.Meta().TempTableType == model.TempTableLocal {
+		// Don't attach, don't lock.
+		return tbl, nil
+	} else if tbl.Meta().TempTableType == model.TempTableGlobal {
+		skipLock = true
+	}
+	if IsAutoCommitTxn(sctx) && sctx.GetSessionVars().StmtCtx.IsReadOnly {
+		return tbl, nil
+	}
+	tableInfo := tbl.Meta()
+	if _, ok := sctx.GetSessionVars().GetRelatedTableForMDL().Load(tableInfo.ID); !ok {
+		if se, ok := is.(*infoschema.SessionExtendedInfoSchema); ok && skipLock {
+			if se.MdlTables == nil {
+				return tbl, nil
+			}
+			if _, ok := se.MdlTables.TableByID(tableInfo.ID); ok {
+				// Already attach.
+				return tbl, nil
+			}
+		}
+		var err error
+		tbl, err = domain.GetDomain(sctx).InfoSchema().TableByName(dbName, tableInfo.Name)
```

Comment:
If here, user drop one table and create one table with same name, if this has problem?